Add a way to configure default search options

thataboy opened the pull request:
Closes https://github.com/zed-industries/zed/issues/4646

```json
// Search options to enable by default when opening new project and buffer searches.
"search": {
  "whole_word": false,
  "case_sensitive": false,
  "include_ignored": false,
  "regex": false
}
```

Release Notes:

- Added `search` settings section to configure default options enabled in buffer and project searches ([#4646](https://github.com/zed-industries/zed/issues/4646))

thataboy committed:
Added default search mode option (text or regex)

{
  // Search mode to use by default. Can be "text" or "regex"
  "search_mode": "text",
}

SomeoneToIgnore reviewed (commented):
I like the PR in general, but I think we might want to

* extract those settings into a dedicated section
* add more defaults, in particular, I crave for "Match case sensitively" default setting

Let's do both here and merge this?

thataboy committed:
Add `search_defaults` setting to set default search options.

{
  // Search options to enable by default.
  "search_defaults": {
    "whole_word": false,
    "case_sensitive": false,
    "include_ignored": false,
    "regex": false
  }
}

thataboy committed:
Fix clippy::redundant_field_names

SomeoneToIgnore commented on assets/settings/default.json:
```diff
@@ -279,6 +279,13 @@
   // If 'search_wrap' is disabled, search result do not wrap around the end of the file.
   "search_wrap": true,
+  // Search options to enable by default.
+  "search_defaults": {
```
What if we call it `search`?
Then we could host more settings if anything comes up in the future.

thataboy replied:
I was going to call it search_options but there's already a struct called SearchOptions. Search is a bit too wide. Maybe search_options and live with the internal and external name not matching!?

SomeoneToIgnore replied:
I am still into just `search` as a key in that json, same as we have `git`, `terminal`, `gutter`, etc.
`"search"` seems to be vacant?

SomeoneToIgnore commented on assets/settings/default.json:
```diff
@@ -279,6 +279,13 @@
   "relative_line_numbers": false,
   // If 'search_wrap' is disabled, search result do not wrap around the end of the file.
   "search_wrap": true,
+  // Search options to enable by default.
```
```suggestion
  // Search options to enable by default when opening new project and buffer searches.
```

SomeoneToIgnore commented on crates/editor/src/editor_settings.rs:
```diff
@@ -155,6 +157,20 @@ pub enum ScrollBeyondLastLine {
     VerticalScrollMargin,
 }
 
+/// Default match options; all false
```
```suggestion
/// Default options for buffer and project search items.
```

SomeoneToIgnore commented on crates/editor/src/editor_settings.rs:
```diff
@@ -249,6 +265,11 @@ pub struct EditorSettingsContent {
 
+    /// Defaults for search options (case sensitive, whole word, include ignored, regex)
+    ///
+    /// Default: all false
```
```suggestion
    /// Default: nothing is enabled
```

SomeoneToIgnore commented on crates/editor/src/editor_settings.rs:
```diff
@@ -249,6 +265,11 @@ pub struct EditorSettingsContent {
     /// Default: true
     pub search_wrap: Option<bool>,
 
+    /// Defaults for search options (case sensitive, whole word, include ignored, regex)
```
```suggestion
    /// Defaults to use when opening a new buffer and project search items.
```

SomeoneToIgnore commented on crates/search/src/buffer_search.rs:
```diff
@@ -504,6 +504,7 @@ impl BufferSearchBar {
         let replacement_editor = cx.new_view(|cx| Editor::single_line(cx));
         cx.subscribe(&replacement_editor, Self::on_replacement_editor_event)
             .detach();
+        let search_options = SearchOptions::from_defaults(&EditorSettings::get_global(cx).search_defaults);
```
Unlike project searches which are instantiated on every action call, buffer search is only created a few times; once for all code-related buffers:

https://github.com/zed-industries/zed/blob/8ee42d37423ac8734595522d680aa186b617ccf3/crates/zed/src/zed.rs#L582


that's why we need to do more here, to reflect the settings changes.
I believe, we can update any of the `deploy` (or `show`/`toggle`/etc. instead?) methods to do the same settings retrieval and an update.

thataboy replied:
Thanks. I'll to look to see where we can update it. You know how I can watch for a change in the settings

SomeoneToIgnore replied:
You could have another subscription like https://github.com/zed-industries/zed/blob/eecff753fbd5ab7eecc886122cd509cb45a97326/crates/outline_panel/src/outline_panel.rs#L656-L662

SomeoneToIgnore reviewed (commented):
Thank you, quite nice that all both search variants are covered by the options.

Please don't get overwhelmed by the PR notes, most of them try to mention "both buffer and project searches" idea in every config-related place — feel free to use something better worded.

Apart from that, the only issue I see is the buffer search not picking up the settings changes.

thataboy committed:
Update assets/settings/default.json

Co-authored-by: Kirill Bulatov <mail4score@gmail.com>

thataboy committed:
Update crates/editor/src/editor_settings.rs

Co-authored-by: Kirill Bulatov <mail4score@gmail.com>

thataboy committed:
Update crates/editor/src/editor_settings.rs

Co-authored-by: Kirill Bulatov <mail4score@gmail.com>

thataboy committed:
Update crates/editor/src/editor_settings.rs

Co-authored-by: Kirill Bulatov <mail4score@gmail.com>

thataboy committed:
Rename `search_defaults` to `search`

thataboy committed:
rename `SearchDefaults` to `SearchSettings`

thataboy committed:
Update buffer search bar when search settings change

thataboy committed:
Update project search bar when search settings change

thataboy commented:
I apologize for the merge conflict. I must be using an older or newer branch.

SomeoneToIgnore reviewed (commented):
I'm a bit uncertain that resetting search toggles for existing buffer/project searches is good: imagine someone opens a project search, adds some filters and gets the results, and then keeps that project search tab somewhere in the pane for a while, opening N tabs afterwards.
Then, if we reset these, project search will retrigger and erase the old results — might not be very pleasant.

Similarly, buffer search filters changing is not very convenient if I set something custom there before altering the settings?

I think we can apply the new ones on any new project/buffer deploy after the settings change instead.

Looks good otherwise, let's fix that, the formatting and merge conflicts and we're good to go.

thataboy committed:
revert live update of project search bar

thataboy committed:
revert live update of buffer search bar

SomeoneToIgnore reviewed (approved):
Thanks, looks good!

There's a merge conflict still, so I cannot merge it yet, alas.

thataboy committed:
Resolve merge conflict

thataboy commented:
Sorry again (I thought you could do it). I resolved the conflict but it'll no longer compile for me with that change so I'm doing it on faith. Hope it's ok.

thataboy commented:
Any thoughts on how to deal with the unit test failing due to `unregistered setting type editor::editor_settings::SearchSettings`. I'm stumped. Thank you for your patience.

SomeoneToIgnore committed:
cargo fmt

SomeoneToIgnore committed:
Fix the tests

SomeoneToIgnore commented:
I had to move the setting registration into `search::init` as it's related to both kinds of searches, and call that `search::init` after workspace initialization in every init method in the failing tests.

You've also missed a `cargo fmt` run as the CI said, that is easy to run locally before pushing.
Tests seem to pass now.

thataboy commented:
Thanks. I just realized last night that if we don't want to live update the search panel, there's no need to observe settings change at all. We can just check in BufferSearchBar.show() if the settings are now different from .default_options. This would've simplified the code.

Should I do another PR or do you want to leave well enough alone? Maybe the current code could be useful in the future.

SomeoneToIgnore commented:
That was my original proposal in https://github.com/zed-industries/zed/pull/17179#discussion_r1746467719 so sure, related follow-up PRs are nice.

urandom commented:
This is a bit late, but shouldn't the new "use_smart_search" be part of this configuration as well? @SomeoneToIgnore

SomeoneToIgnore commented:
Moving that one would be a breaking change, so it's a good thing that we did not touch it during this PR.

We have https://github.com/zed-industries/zed/issues/17742 now, so at least we should not forget to move it when the time comes.
